modules/dds/image_loader_dds.cpp

```diff
@@ -0,0 +1,421 @@
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "image_loader_dds.h"
+
+#include "core/os/os.h"
+
+#include "core/io/file_access.h"
+#include "core/io/file_access_memory.h"
+#include <string.h>
+
+#define PF_FOURCC(s) ((uint32_t)(((s)[3] << 24U) | ((s)[2] << 16U) | ((s)[1] << 8U) | ((s)[0])))
+
+// Reference: https://docs.microsoft.com/en-us/windows/win32/direct3ddds/dds-header
+
+enum {
+	DDS_MAGIC = 0x20534444,
+	DDSD_PITCH = 0x00000008,
+	DDSD_LINEARSIZE = 0x00080000,
+	DDSD_MIPMAPCOUNT = 0x00020000,
+	DDPF_FOURCC = 0x00000004,
+	DDPF_ALPHAPIXELS = 0x00000001,
+	DDPF_INDEXED = 0x00000020,
+	DDPF_RGB = 0x00000040,
+};
+
+enum DDSFormat {
+	DDS_DXT1,
+	DDS_DXT3,
+	DDS_DXT5,
+	DDS_ATI1,
+	DDS_ATI2,
+	DDS_A2XY,
+	DDS_BGRA8,
+	DDS_BGR8,
+	DDS_RGBA8, //flipped in dds
+	DDS_RGB8, //flipped in dds
+	DDS_BGR5A1,
+	DDS_BGR565,
+	DDS_BGR10A2,
+	DDS_INDEXED,
+	DDS_LUMINANCE,
+	DDS_LUMINANCE_ALPHA,
+	DDS_MAX
+};
+
+struct DDSFormatInfo {
+	const char *name = nullptr;
+	bool compressed = false;
+	bool palette = false;
+	uint32_t divisor = 0;
+	uint32_t block_size = 0;
+	Image::Format format = Image::Format::FORMAT_BPTC_RGBA;
+};
+
+static const DDSFormatInfo dds_format_info[DDS_MAX] = {
+	{ "DXT1/BC1", true, false, 4, 8, Image::FORMAT_DXT1 },
+	{ "DXT3/BC2", true, false, 4, 16, Image::FORMAT_DXT3 },
+	{ "DXT5/BC3", true, false, 4, 16, Image::FORMAT_DXT5 },
+	{ "ATI1/BC4", true, false, 4, 8, Image::FORMAT_RGTC_R },
+	{ "ATI2/3DC/BC5", true, false, 4, 16, Image::FORMAT_RGTC_RG },
+	{ "A2XY/DXN/BC5", true, false, 4, 16, Image::FORMAT_RGTC_RG },
+	{ "BGRA8", false, false, 1, 4, Image::FORMAT_RGBA8 },
+	{ "BGR8", false, false, 1, 3, Image::FORMAT_RGB8 },
+	{ "RGBA8", false, false, 1, 4, Image::FORMAT_RGBA8 },
+	{ "RGB8", false, false, 1, 3, Image::FORMAT_RGB8 },
+	{ "BGR5A1", false, false, 1, 2, Image::FORMAT_RGBA8 },
+	{ "BGR565", false, false, 1, 2, Image::FORMAT_RGB8 },
+	{ "BGR10A2", false, false, 1, 4, Image::FORMAT_RGBA8 },
+	{ "GRAYSCALE", false, false, 1, 1, Image::FORMAT_L8 },
+	{ "GRAYSCALE_ALPHA", false, false, 1, 2, Image::FORMAT_LA8 }
+};
+
+static Ref<Image> dds_load_image_from_buffer(const uint8_t *p_buffer, int p_buffer_len) {
```

Comment:
I'm wondering if this function name should start with an underscore. Even in the future it shouldn't conflict with anything else, but maybe it would be better for the convention. Or, because it is directly bound to `Image::_dds_mem_loader_func`, maybe it should be just named `_dds_mem_loader_func`?

Reply:
I've renamed this function to match the convention.